discord/interaction.go

```diff
@@ -0,0 +1,129 @@
+package discord
+
+import (
+	"bytes"
+
+	"github.com/diamondburned/arikawa/v3/utils/json"
+)
+
+// https://discord.com/developers/docs/topics/gateway#interactions
+type Interaction struct {
+	ID        InteractionID   `json:"id"`
+	AppID     AppID           `json:"application_id"`
+	Type      InteractionType `json:"type"`
+	Data      InteractionData `json:"data,omitempty"`
+	ChannelID ChannelID       `json:"channel_id,omitempty"`
+	Token     string          `json:"token"`
+	Version   int             `json:"version"`
+	Message   *Message        `json:"message"`
+
+	// Member is only present if this came from a guild.
+	Member  *Member `json:"member,omitempty"`
+	GuildID GuildID `json:"guild_id,omitempty"`
+
+	// User is only present if this didn't come from a guild.
+	User *User `json:"user,omitempty"`
+}
+
+func (i *Interaction) UnmarshalJSON(p []byte) error {
+	type interaction Interaction
+	v := struct {
+		Data json.Raw `json:"data,omitempty"`
+		*interaction
+	}{interaction: (*interaction)(i)}
+	if err := json.Unmarshal(p, &v); err != nil {
+		return err
+	}
+
+	switch v.Type {
+	case PingInteraction:
+		return nil
+	case ComponentInteraction:
+		i.Data = &ComponentInteractionData{}
+	case CommandInteraction:
+		i.Data = &CommandInteractionData{}
+	default:
+		i.Data = &UnknownInteractionData{typ: v.Type}
+	}
+
+	return json.Unmarshal(v.Data, i.Data)
+}
+
+type InteractionType uint
+
+const (
+	PingInteraction InteractionType = iota + 1
+	CommandInteraction
+	ComponentInteraction
+)
+
+// InteractionData holds the data of an interaction.
+// Type assertions should be made on InteractionData to
+// access the underlying data.
+type InteractionData interface {
+	Type() InteractionType
+}
+
+type ComponentInteractionData struct {
+	CustomID      string        `json:"custom_id"`
+	ComponentType ComponentType `json:"component_type"`
+	Values        []string      `json:"values"`
+}
+
+func (c ComponentInteractionData) Type() InteractionType {
+	return ComponentInteraction
+}
+
+type CommandInteractionData struct {
+	ID      CommandID           `json:"id"`
+	Name    string              `json:"name"`
+	Options []InteractionOption `json:"options"`
+}
+
+func (c CommandInteractionData) Type() InteractionType {
+	return CommandInteraction
+}
+
+type UnknownInteractionData struct {
+	json.Raw
+	typ InteractionType
+}
+
+func (u UnknownInteractionData) Type() InteractionType {
+	return u.typ
+}
+
+type InteractionOption struct {
+	Name    string              `json:"name"`
+	Value   json.Raw            `json:"value"`
+	Options []InteractionOption `json:"options"`
+}
+
+func (o InteractionOption) String() string {
```

Comment:
Clarify that `String()` will never fail and will always print `Value` as either
a valid string or a raw value of other type.